Src/Base/AMReX_FabArray.H

```diff
@@ -1561,56 +1573,99 @@ FabArray<FAB>::AllocFabs (const FabFactory<FAB>& factory, Arena* ar,
 
 template <class FAB>
 void
-FabArray<FAB>::setFab (int  boxno,
-                       FAB* elem)
+FabArray<FAB>::setFab_assert (int K, FAB const& fab) const
+{
+    amrex::ignore_unused(K,fab);
+    AMREX_ASSERT(n_comp == fab.nComp());
+    AMREX_ASSERT(boxarray.size() > 0);
+    AMREX_ASSERT(fab.box() == fabbox(K));
+    AMREX_ASSERT(distributionMap[K] == ParallelDescriptor::MyProc());
+}
+
+template <class FAB>
+void
+FabArray<FAB>::setFab (int boxno, std::unique_ptr<FAB> elem)
 {
-    //
-    // Must check it is of the proper size.
-    //
     if (n_comp == 0) {
         n_comp = elem->nComp();
     }
 
-    BL_ASSERT(n_comp == elem->nComp());
-    BL_ASSERT(boxarray.size() > 0);
-    BL_ASSERT(elem->box() == fabbox(boxno));
-    BL_ASSERT(!this->defined(boxno));
-    BL_ASSERT(distributionMap[boxno] == ParallelDescriptor::MyProc());
+    setFab_assert(boxno, *elem);
 
     if (m_fabs_v.size() == 0) {
-      m_fabs_v.resize(indexArray.size(),nullptr);
+        m_fabs_v.resize(indexArray.size(),nullptr);
     }
 
     const int li = localindex(boxno);
-    m_fabs_v[li] = elem;
+    if (m_fabs_v[li]) {
+        m_factory->destroy(m_fabs_v[li]);
+    }
+    m_fabs_v[li] = elem.release();
 }
 
 template <class FAB>
+template <class F, std::enable_if_t<std::is_move_constructible<F>::value,int> >
 void
-FabArray<FAB>::setFab (const MFIter& mfi,
-                       FAB* elem, bool assertion)
+FabArray<FAB>::setFab (int boxno, FAB&& elem)
 {
-    if (assertion)
-    {
-        //
-        // Must check it is of the proper size.
-        //
-        if (n_comp == 0)
-            n_comp = elem->nComp();
+    if (n_comp == 0) {
+        n_comp = elem.nComp();
+    }
+
+    setFab_assert(boxno, elem);
+
+    if (m_fabs_v.size() == 0) {
+        m_fabs_v.resize(indexArray.size(),nullptr);
+    }
+
+    const int li = localindex(boxno);
+    if (m_fabs_v[li]) {
+        m_factory->destroy(m_fabs_v[li]);
+    }
+    m_fabs_v[li] = new FAB(std::move(elem));
+}
 
```

Comment:
Is it possible to collapse these down into fewer copies of the same function? e.g. MFIter version calls the int version?

Looks like the problem is mfi.index() and mfi.LocalIndex(). Would it make sense to have a setFab_doit in this case, or too much hassle?

Reply:
We could get the global index with `MFIter::index()` and call the int version, not the other way around.  But the MFIter version is more efficient than the int version.